Blog post for moving data between JBOD disks using Cruise Control (#469)
strimzi/strimzi.github.io: _posts/2024-12-19-moving-data-between-JBOD-disks-using-cruise-control.md

---
layout: post
title: "Moving data between the JBOD disks using Cruise Control"
date: 2024-12-19
author: shubham_rawat
---

> Apache Kafka is a platform which provides durability and fault tolerance by storing messages on disks and JBOD storage is one of the storage configuration types supported by Kafka.

Reviewer: Kafka supports multiple disks. JBOD storage configuration type is more of a Strimzi thing. That is how we call it in the API. I would either reword it or change it from Kafka to Strimzi?

The JBOD data storage configuration allows Kafka brokers to make use of multiple disks.
Using JBOD storage, you can increase the data storage capacity for Kafka nodes, which can further lead to performance improvements.
In case you plan to remove a disk, and it contains some partition replicas, then you need to make sure that data is safely moved to some other disks.

> If the data is not removed from the disk, and it is removed then potential data loss can happen.

Reviewer: Maybe ...

```suggestion
If the data is not removed from the disk, and the disk is removed then potential data loss can happen.
```

> Currently, moving data between the JBOD disks is done using the `kafka-reassign-partitions.sh` tool which is not very user-friendly, therefore in Strimzi 0.45.0 we are introducing the ability to move data between the JBOD disks using Cruise Control

Reviewer: ```suggestion
Moving data between the JBOD disks can be done using the `kafka-reassign-partitions.sh` tool which is not very user-friendly, therefore in Strimzi 0.45.0 we are introducing the ability to move data between the JBOD disks using Cruise Control
```

> ## Cruise Control to move data between JBOD disks

Reviewer: If you move all your headers one level down, the webpage would look nicer ... i.e. `##` -> `###`, `###` -> `####` etc.

> This feature will allow you to move the data between the JBOD disks using the `KafkaRebalance` custom resource that we have in Strimzi.

Reviewer: `Between the JBOd disks` sounds not right here. That is what the intrabroker rebalancing does which we have for some time already. You are moving all data from one disk to another disks.

> This feature makes use of the `remove-disks` endpoint of Cruise Control that triggers a rebalancing operation which moves replicas, starting with the largest and proceeding to the smallest, to the remaining disks. 

Reviewer: ```suggestion
This feature makes use of the `remove-disks` endpoint of Cruise Control that triggers a rebalancing operation which moves all replicas, starting with the largest and proceeding to the smallest, to the remaining disks. 
```

## Setting up the environment

Let's set up a cluster to work through an example demonstrating this feature.

> To get the Kafka cluster up and running, we will first have to install the Strimzi Cluster Operator and then deploy the `Kafka` resource.

Reviewer 2: It's not going to be just `Kafka` but also `KafkaNodePool`.

You can install the Cluster Operator with any installation method you prefer.

> For example by following the [Stimzi Quickstart Guide](https://strimzi.io/quickstarts/).

Reviewer: Should you point to the docs instead? The quickstart deploys a Kafka cluster as well.

Reviewer 3: This example might benefit from a brief explanation here of each of the resources being used. 
Something like:

> We use a `Kafka` resource and `KafkaNodePool` resources to create a Kafka cluster.
> Then, we create a `KafkaRebalance` resource in `remove-disks` mode, specifying the brokers and volume IDs for partition reassignment. 
> After generating the optimization proposal, we approve it to move the data.

Then we will deploy the Kafka cluster with JBOD storage and Cruise Control enabled.
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  name: pool-a
  labels:
    strimzi.io/cluster: my-cluster
spec:
  replicas: 3
  roles:
    - broker
  storage:
    type: jbod
    volumes:
      - id: 0
        type: persistent-claim
        size: 100Gi
        deleteClaim: false
      - id: 1
        type: persistent-claim
        size: 100Gi
        deleteClaim: false
      - id: 2
        type: persistent-claim
        size: 100Gi
        deleteClaim: false
---

apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  name: pool-b
  labels:
    strimzi.io/cluster: my-cluster
spec:
  replicas: 3
  roles:

>     - broker

Reviewer: Should this be controller?

Reviewer 2: I guess that based on the rest of the YAMLs, @ShubhamRwt is making the example with a ZooKeeper based cluster but I totally agree on using KRaft only from now on.

Author: Yes, I will update this with a KRaft cluster

  storage:
    type: jbod
    volumes:
      - id: 0
        type: persistent-claim
        size: 200Gi
        deleteClaim: false
---

apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
  annotations:

>     strimzi.io/node-pools: enabled

Reviewer 2: missing annotation to specify it's a KRaft based cluster and so remove the ZooKeeper section as well.

spec:
  kafka:
    version: 3.9.0
    listeners:
      - name: plain
        port: 9092
        type: internal
        tls: false
      - name: tls
        port: 9093
        type: internal
        tls: true
    config:
      offsets.topic.replication.factor: 3
      transaction.state.log.replication.factor: 3
      transaction.state.log.min.isr: 2
      default.replication.factor: 3
      min.insync.replicas: 2
      inter.broker.protocol.version: "3.9"
  zookeeper:
    replicas: 3
    storage:
      type: persistent-claim
      size: 100Gi

>       deleteClaim: false

Reviewer: Use KRaft please.

              "partition": "my-topic-3",
              "size": 0,
              "offsetLag": 0,
              "isFuture": false
            }
          ],
          "error": null,
          "logDir": "/var/lib/kafka/data-0/kafka-log0"
        },
        {
          "partitions": [
            {
              "partition": "my-topic-4",
              "size": 0,
              "offsetLag": 0,
              "isFuture": false
            },
            {
              "partition": "my-topic-1",
              "size": 0,
              "offsetLag": 0,
              "isFuture": false
            }
          ],
          "error": null,
          "logDir": "/var/lib/kafka/data-1/kafka-log0"
        }
      ]
    }
  ]
}
```

Now lets try to move the data of volume 1 and volume 2 to volume 0, present on broker with ID 0. For doing that let's create a `KafkaRebalance` resource with `remove-disks` mode.
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
  labels:
    strimzi.io/cluster: my-cluster
# no goals specified, using the default goals from the Cruise Control configuration
spec:
  mode: remove-disks
  moveReplicasOffVolumes:
    - brokerId: 0

>       volumeIds: [1, 2]

Reviewer: How do you know what will be broker 0? You should probably also clean it from all brokers from the node pool and take the example till the end to remove the disks?

Author: @scholzj  Sorry, I didn't understand what you meant by `How do you know what will be broker 0`.  I was just showing how we can move all the data from the two volumes to some other volume

Reviewer: How do you know the ID 0 will not be a broker form the second node pool with only one volume? The YAML you use here does not give you deterministic order in which the node IDs will be assigned.

Author: @scholzj Hi, before I push my changes if I understood the above suggestion correctly -> Iam now using a Kraft cluster with broker having 3 disks and then removing the 3rd disk from all the brokers(not the controllers)  and taking the ecample till the end where we remove the disks

Reviewer: The point of this comment was mainly that without the node ID annotations, ou do not know if nodes 0, 1 and 2 will be brokers or controllers.

```

> Now let’s wait for the `KafkaRebalance` resource to move to `ProposalReady` state. You can check the rebalance summary by running the following command once the proposal is ready:

Author: @scholzj for eg. here. I see the text in green when it should be white as I have above for normal text

Reviewer: You mean in the GitHub review window? Or on the blog preview?

Author: In the GitHub review window

Reviewer: So, from my experience, that is quite easily derailed by some code examples etc. So I would not make too muhc out of it as long as the preview looks good.

Author: Okay, thanks for clearing the doubt. I though I have some wring formatting or something which is causing this

```shell
kubectl get kafkarebalance my-rebalance -n myproject -o yaml
```

> and you should be able to get an output like this:

Reviewer 3: ```suggestion
And you should be able to get an output like this:
```

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  creationTimestamp: "2024-11-13T06:55:41Z"
  generation: 1
  labels:
    strimzi.io/cluster: my-cluster
  name: my-rebalance
  namespace: myproject
  resourceVersion: "5354"
  uid: 7aacb9b0-489b-4d73-bfd7-bafbac676109
spec:
  mode: remove-disks
  moveReplicasOffVolumes:
  - brokerId: 0
    volumeIds:
    - 1
    - 2
status:
  conditions:
  - lastTransitionTime: "2024-11-13T06:55:42.217794891Z"
    status: "True"
    type: ProposalReady
  observedGeneration: 1
  optimizationResult:
    afterBeforeLoadConfigMap: my-rebalance
    dataToMoveMB: 0
    excludedBrokersForLeadership: []
    excludedBrokersForReplicaMove: []
    excludedTopics: []
    intraBrokerDataToMoveMB: 0
    monitoredPartitionsPercentage: 100
    numIntraBrokerReplicaMovements: 26
    numLeaderMovements: 0
    numReplicaMovements: 0
    onDemandBalancednessScoreAfter: 100
    onDemandBalancednessScoreBefore: 0
    provisionRecommendation: ""
    provisionStatus: UNDECIDED
    recentWindows: 1
  sessionId: 24537b9c-a315-4715-8e86-01481e914771
```

> Now you can use the `approve` annotation to apply the generated proposal.

Reviewer 3: do we want to mention auto-approval?

> Once the rebalance is complete, you can then use the `kafka-log-dirs.sh` tool again to check if the data moved or not:

Reviewer 3: ```suggestion
After the rebalance is complete, use the `kafka-log-dirs.sh` tool again to verify that the data has been moved.
```

            {
              "partition": "my-topic-4",
              "size": 0,
              "offsetLag": 0,
              "isFuture": false
            },
            {
              "partition": "my-topic-5",
              "size": 0,
              "offsetLag": 0,
              "isFuture": false
            },
            {
              "partition": "my-topic-0",
              "size": 0,
              "offsetLag": 0,
              "isFuture": false
            },
            {
              "partition": "my-topic-1",
              "size": 0,
              "offsetLag": 0,
              "isFuture": false
            },
            {
              "partition": "my-topic-2",
              "size": 0,
              "offsetLag": 0,
              "isFuture": false
            },
            {
              "partition": "my-topic-3",
              "size": 0,
              "offsetLag": 0,
              "isFuture": false
            }
          ],
          "error": null,
          "logDir": "/var/lib/kafka/data-0/kafka-log0"
        },
        {
          "partitions": [],
          "error": null,
          "logDir": "/var/lib/kafka/data-1/kafka-log0"
        }
      ]
```

> As you can see in the output, the data has moved from volume 1 and volume 2, and they now don’t have any partition replicas on them. All the partition replicas are now moved to volume 0 on broker with ID 0.

Reviewer 3: ```suggestion
As shown in the output, the data has been moved from volumes 1 and 2, which no longer contain any partition replicas. 
All partition replicas are now located on volume 0 of the broker with ID 0.
```

### Additional notes

> 1. This feature only works if JBOD storage is enabled

Author: Here also I see the text in blue while these are just normal text and should be in white If if Iam not wrong?

> 1. This feature only works if JBOD storage is enabled

Reviewer: It only works when multiple disks are used, not when JBOD is enabled but has 1 disk, or?

> 1. This feature only works if JBOD storage is enabled

Reviewer 3: ```suggestion
1. This feature only works if JBOD storage is enabled and multiple disks are used.
```

> 2. Make sure you have more than one volume per broker else you will be prompted of not having enough volumes to move the data to.

Reviewer 3: ```suggestion
2. Make sure you have more than one volume per broker else you will be prompted for not having enough volumes to move the data to.
```

> 3. This endpoint does not provide `before` load since upstream Cruise Control project does not support `verbose` with this endpoint so the `loadmap` generated should only have `afterLoad` information.

Reviewer 3: ```suggestion
3. The optimization proposal does not show the load before optimization, it only shows the load after optimization. 
This is because in upstream Cruise Control the verbose tag is not enabled with the `remove_disks` endpoint.
```

> ## What's next

Reviewer: You should probably list the risks / incomplete parts as well ... in paritcular, new partition replicas might be scheduled to the disks between cleaning them up with cruise control and removing them which might lead to data loss again.

We hope this blog post has provided you with a clear understanding of how you can use the `KafkaRebalance` custom resource in `remove-disks` to easily move the data between the JBOD disks. 

> If you get stuck on any step or have any doubts, you can have read about this in or documentation on [Using Cruise Control to reassign parititon on JBOD disk](https://strimzi.io/docs/operators/latest/deploying#proc-cruise-control-moving-data-str)

Reviewer 3: ```suggestion
If you encounter any issues or want to know more, refer to our documentation on [Using Cruise Control to reassign partitions on JBOD disks](https://strimzi.io/docs/operators/latest/deploying#proc-cruise-control-moving-data-str)
```
